### PR TITLE
[IMP] Allow addons to fully support multiple Odoo versions

### PR DESCRIPTION
**Pull request by yajo**
Sorry, I had this floating on my mind and couldn't wait anymore to publish it :sweat_smile:. This is just a WIP for now, so you can have an example implementation, and if it gets approved, I can start working with tests and docs.

In order to make possible to implement a different workflow on downstream addon maintenance teams, as explained in https://github.com/OCA/maintainer-tools/issues/334, Odoo needs to support version-specific code.

This can be currently done in Python by checking `odoo.release.version_info` and in JS by using the `/web/webclient/version_info` controller. However, when dealing with data files, we were out of luck.

With this patch:

- A new cached helper is found in JS sessions, that lazy-caches version info once per session, which will make version-specific assertions a pleasure in JS.
- A new attribute `versions` is supported in `<odoo>`, `<openerp>` and `<data>` elements within XML data files, that allows the programmer to specify a version spec that makes that data section be omitted when Odoo version doesn't match it.
- A new `overrides` key is supported in the `__manifest__.py` file, which has the form of a dict, where keys are version specs, and values are dicts that will directly override the addon base manifest.

So, for instance, to declare an addon to be available only until version 13.0 of Odoo, put this in the manifest:

```python
{
  ...
  "installable": False,
  "overrides": {
    ":13": {
      "installable": True,
    }
  }
}
```

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr
@Tecnativa @sbidoul @mart-e @tde-banana-odoo @jbeficent

### CODE REVIEW

**Comment by Yajo** [comment]
The discussion on how to use this downstream goes in https://github.com/OCA/maintainer-tools/issues/334; but before we decide the best way to use it, the 1st requirement is that it's actually *possible* to do it, which is this PR's purpose.

**Comment by odony** [comment]
We really don't want to go in that direction. We manage official addons as a consistent codebase, branch-based, and not a loosely coupled distribution of bits and pieces. It's been discussed at length many times in the past, every time someone comes up with another versioning and dependency scheme. 

As we don't need or want such a mechanism the official distribution, it makes no sense to encourage it for third-party ones.

**Comment by Yajo** [comment]
I have to ask you to please reconsider this subject. Let me explain a little bit of the reasoning behind this.

The first thing you have to understand is that **I'm absolutely not asking you (or anybody else) here to change the release system**. I agree that Odoo is a fully-featured complete product, and that your release system works really fine _for you_, as the product owner and designer. Having stable releases, once per year, as you are doing now, gives you a very good balance between maintainability and innovation.

However, what happens with Odoo SA and what happens in the downstream of the product (a.k.a. the community, a.k.a. your partners and @OCA) is dramatically different. See this simple table:

| Subject | Odoo SA | Partners | OCA |
|----------------------------------------------------------------------------|----------------------------------------------------------------|---------------------------|---------------------------|
| Release schedule | Fixed, 1 stable release per year | Depends on customer needs | Depends on voluntary work |
| Release support | Fixed, 3 latest stable releases | Depends on customer needs | Virtually endless |
| Amount of addons maintained for Odoo 8.0 | 0 | Unknown | 1254 |
| Amount of addons maintained for Odoo 9.0 | 275 (CE) | Unknown | 777 |
| Amount of addons maintained for Odoo 10.0 | 267 (CE) | Unknown | 937 |
| Amount of addons maintained for Odoo 11.0 | 275 (CE) | Unknown | 358 |
| Amount of addons presumably available for Odoo 12.0 since day 1 of release | 275 (CE) | 0 | 0 |
| Total addons/version maintained (8-11) | 817 (CE) | Unkown | 3326 |
| Multi-version patch support | Somebody is paid to forward-port patches in supported branches | Depends on customer needs | Depends on voluntary work |
| **What's your product?** | **Odoo CE and EE** | **Your customer's database** | **Addons, mostly** |

So, as you can notice, while the world inside Odoo SA's office follows very simple rules, the outside world is not so simple. This produces some difficult situations:

- Your customer might have a bug from some OCA addon in version 8.0, you have fixed that version, and when you upgrade him, you lose the patch and get the bug in again.
- Same, but with features.
- Some customers are forced to migrate to newer versions just to get new features from community addons (unrelated to the underlying Odoo version).
- Literally **thousands** of human work hours are wasted every year, out there in the community and partner groups, to get addons from older versions to newer versions, *when 90% of the time, the addon would work as-is, unchanged*.
- If OCA still needs to forward-port 579 (!!!) addons to version 11.0 to match what it has for v10, when version 12 is released, that maintenance window doesn't get *displaced* (as happens to Odoo SA), it gets *increased*.

So, the maintenance burden produced by the current release system is:

- For Odoo SA: mostly always the same.
- For partners: many of them just will stick with v8 forever; others try to keep the pace, but since that's mostly impossible nowadays, they tend to use some LTS-approach: from v8 to v11; from v11 to v15; etc.
- For OCA: exponentially, with each new release, **it has to maintain the same amount of code it has, but in a new, completely unrelated, branch (which is empty to start, BTW)**

So, summarizing, while the lack of this feature means nothing for Odoo SA, **it's placing the community in an exponentially unmaintainable position**.

That's why the last row in the above table is bold: your current release system is really good... for Odoo SA, which produces the Odoo product. But **for downstream partners and community, whose product is (almost) a completely different thing, it's devastating.**

**For downstream, Odoo is like our operating system: that piece of software where you install the software you develop.** Or, at least, that's how it should be. :stuck_out_tongue_winking_eye: 

Many (really: many many) partners out there *fear* new releases. They are struggling to keep up with Odoo's pace, and a new release means new maintenance burden, and every year it's the same... A common complaint I hear is: *Odoo advances too fast!*

So, it's the time to **stop, look behind, look ahead, and make ourselves some questions**:

- Is it really the quick Odoo development a problem? That should never be true!
- Does it make sense that we, downstream, have to rewrite code every year? Nope.
- And if, above that, we have to rewrite our commit history? Nope!
- And if, above that, we have to review again some code that was already in? Definitely nope!
- Do I know any other open source community that keeps doing that? Nope.
- Is that DRY? Nope.
- Is that maintainable (downstream)? Nope.
- Could that be the root of the "OCA is too hard" problem? Most likely, yes.
- Does Odoo maintain separate branches for installing in Windows, in Debian, in CentOS, etc.? No! That'd be crazy! Just some `if`s here and there, and you go.
- So, if Odoo is like my OS, should I maintain a different branch for installing my addons in Odoo 12, 13, 14...? Nope; for **exactly the same reason**.
- Are the community and partners important for Odoo SA? I know we are :heart: 
- Is it reasonable that Odoo SA empowers downstream with tools that can potentially save thousands of hours of work? Absolutely!
- Should I spend so much time migrating addons, given Odoo is almost always almost completely backwards compatible when introducing new features? Nope, most of the time.
- What should I care migrating? Most likely **your customer's database, not your code**.
- What could I do with all that free time I'd have if I were able to forget about all of this nonsense? Not sure... Actually developing something new? Vacation instead? :sun_with_face: 
- And the most important one: **can we fix the situation?** Yes! Just merge this pull request! (Well, at least once I develop some tests).

Then, what could potentially be our benefit of you merging this 70-line patch?

- We could maintain a single branch of code.
- Bug fixes shared across versions.
- New features shared accross versions.
- Less dramatic database migrations.
- More time to invest in database migrations.
- We could declare one point in history where some odoo version is no longer officially maintained (for instance, when Odoo releases a not-so-backwards-compatible version, like it was 8.0, or deprecating a version that is more than 4 years old... it's up to downstream!) by just releasing its own branch, as of today, and keeping master branch only with modern Odoo versions support.

However, I said above that I'm asking nobody to change that release schedule here (I want to discuss this within OCA when it's **actually possible**, but all I explained to you here is so you get a grasp of how's the world through the other side of the glass). But leaving this power in the hands of downstream, will make us again embrace Odoo and its fast development pace! :heart: 

Thanks!

---

Just noting that all of this is only my own personal opinion, based on my own personal experience.

**Comment by odony** [comment]
Thanks for the dramatic read ;-) We've heard that rationale many times, to support similar demands.

Simply put, we don't want to encourage version-specific code in Odoo. You're free to do it if you like "downstream", but we won't bless it with an official API and version checks. Especially if we won't be using it. This sort of discrepancy only causes greater frustrations in the long run, due to different expectations.

We consider that sharing codebases across versions encourages bad practices: spaghetti code, code rot, technical debt, etc. Reviewing modules for each new major version is a sound practice that also gives an opportunity of refactoring and cleaning up old code, instead of increasing the debt.
We want to encourage this periodical cleanup. The minimal amount of work that is necessary is often small, but going through it explicitly also gives a chance of noticing more subtle compatibility problems.

**Comment by Yajo** [comment]
> You're free to do it if you like "downstream"

It's not nice to say that while I just explained very clearly that it is *impossible* unless some little additions are made upstream. I can reduce even more the diff, if that's the problem.

It's true that code quality is important, but if we have more time for other things, we can actually spend our time in *real* quality improvements, such as automated unit tests that provide good coverage and ensure the addon works as expected. Addon migration should be more tests and less rewriting.

Simply put, you're unilaterally deciding how all your community should organize their workflow, enforcing what you think are good practices but completely prohibiting other widely recognized good practices (DRY, if something works don't fix it, do not refactor unless needed, better improve coverage than review code) that are more important for other people (that casually happen to be your community and partners, which should be important to you IMHO).

I hope you reconsider this.

**Comment by antonylesuisse** [comment]
Stop wasting your time beating a dead horse. It's a NO ! Our policy was explained thoughtfully by Olivier. You may disagree, but please respect our decision. Thanks.

**Comment by rvalyi** [comment]
@Yago and others: I just came to this thread randomly today. Well at Akretion what we do to decrease this maintenance burden is we sometimes skip new Odoo versions. For instance we skipped version 9 (because it was quite buggy on accouting in fact) and we also skipped version 11 just to simplify the maintenance (while version 11 was arguably better than version 10 but may be without enough benefit to afford the cost of eating a new version). We may skip version 13 too unless we indeed come with a multi-version solution that works somewhat for most of the addons are specialized in. Our customers often prefer zero bug and the horizontal innovation we and the OCA do in the extra addons than being on the bleeding edge at the cost of regressions and endless port work.

From a business prospective Akretion did fine with doing that. However, mostly in Brazil, some official partners who were pressured by Odoo SA to work on the very last version were unable to properly migrate the localization, did some quick dirty non OCA forks and consequently failed their projects and failed themselves, flaming the Odoo product image along the way, something that could have been avoided if Odoo SA paid more attention to your point.
